c/extras/aggregator.cc

```diff
@@ -0,0 +1,351 @@
+      case LType::BOOL:
+      case LType::INT:
+      case LType::REAL:    cols[i] = dt_in->columns[i]->cast(SType::FLOAT64); break;
+      default:                cols[i] = dt_in->columns[i]->shallowcopy();
+    }
+  }
+
+  cols[dt_in->ncols] = Column::new_data_column(SType::INT32, dt_in->nrows);
+  cols[dt_in->ncols + 1] = nullptr;
+  dt_out = DataTablePtr(new DataTable(cols));
+
+  return nullptr;
+}
+
+
+DataTablePtr Aggregator::aggregate(int32_t n_bins, int32_t nx_bins, int32_t ny_bins, int32_t max_dimensions, unsigned int seed) {
+  switch (dt_out->ncols) {
+    case 2:  aggregate_1d(n_bins); break;
+    case 3:  aggregate_2d(nx_bins, ny_bins); break;
+    default: aggregate_nd(max_dimensions, seed);
+  }
+
+  return std::move(dt_out);
+}
+
+
+void Aggregator::aggregate_1d(int32_t n_bins) {
+  LType ltype = info(dt_out->columns[0]->stype()).ltype();
+
+  switch (ltype) {
+    case LType::BOOL:
+    case LType::INT:
+    case LType::REAL:   aggregate_1d_continuous(n_bins); break;
+    case LType::STRING: aggregate_1d_categorical(/*n_bins*/); break;
+    default:            throw ValueError() << "Datatype is not supported";
+  }
+}
+
+
+void Aggregator::aggregate_2d(int32_t nx_bins, int32_t ny_bins) {
+  LType ltype0 = info(dt_out->columns[0]->stype()).ltype();
+  LType ltype1 = info(dt_out->columns[1]->stype()).ltype();
+
+  switch (ltype0) {
+    case LType::BOOL:
+    case LType::INT:
+    case LType::REAL:   {
+                           switch (ltype1) {
+                             case LType::BOOL:
+                             case LType::INT:
+                             case LType::REAL:   aggregate_2d_continuous(nx_bins, ny_bins); break;
+                             case LType::STRING: aggregate_2d_mixed(0, nx_bins/*, ny_bins*/); break;
+                             default:            throw ValueError() << "Datatype is not supported";
+                           }
+                        }
+                        break;
+
+    case LType::STRING: {
+                           switch (ltype1) {
+                             case LType::BOOL:
+                             case LType::INT:
+                             case LType::REAL:   aggregate_2d_mixed(1, nx_bins/*, ny_bins*/); break;
+                             case LType::STRING: aggregate_2d_categorical(/*nx_bins, ny_bins*/); break;
+                             default:            throw ValueError() << "Datatype is not supported";
+                           }
+                        }
+                        break;
+
+    default:            throw ValueError() << "Datatype is not supported";
+  }
+}
+
+
+void Aggregator::aggregate_1d_continuous(int32_t n_bins) {
+  RealColumn<double>* c0 = static_cast<RealColumn<double>*>(dt_out->columns[0]);
+  const double* d_c0 = static_cast<const double*>(dt_out->columns[0]->data());
+  int32_t* d_c1 = static_cast<int32_t*>(dt_out->columns[1]->data_w());
+
+//TODO: handle the case when the column is constant, i.e. min = max
```

Comment:
Please indent comments (here and below) at the same level as the surrounding code, otherwise it messes up with code-folding in my editor.